### PR TITLE
29 implement forgot password functionality

### LINKED ISSUE #29: Implement "Forgot Password" Functionality
We need to add the "Forgot Password" functionality to the application. This feature should allow users to reset their password in case they forget it. Below are the key requirements and tasks for this implementation:

Acceptance criteria:
- UI Component:
   - Clicking the "Forgot Password" link should open a form where users enter their email to receive a reset link.

Figma Design:
https://www.figma.com/design/d2Sezz6VNO6hbnobggGTRq/Forgot-Password-Page?node-id=0-1&t=53YxTrSKyXJP3ALG-1

image

- Backend Logic:
  - Create a route to handle the password reset request (e.g., /forgot-password).
  - Validate the email input to ensure it exists in the database.
  - Generate a secure, time-limited reset token.
  - Send an email to the user with the reset link containing the token.

- Email System:
  - Use a reliable email service (e.g., SendGrid, Nodemailer) to send the reset link.
  - Ensure the reset link is in the format /reset-password?token=<generated_token>.

- Reset Password Page:
  - Create a page where users can set a new password using the reset token.
  - Validate the token on the server to ensure it is valid and not expired.
  - Allow users to submit a new password and confirm it.
  - Show proper messages that password change is successful
  
Figma Design:
https://www.figma.com/design/ulbE7ij6TJuekULrBsHgzq/Reset-password?node-id=0-1&t=wZTmUAzwQma4iiF7-1
image

- Password Reset Logic:
  - Upon receiving the new password, update the user’s password securely (e.g., use Bcrypt hashing).
  - Invalidate the reset token after use.

- Testing:
  - Ensure automated tests cover the following:
  - UI interactions for the forgot password flow.
  - Backend validation and error handling (invalid email, expired token, etc.).
  - Successful password reset scenario.

### PR DESCRIPTION
**Pull request by oam67**
Fixes #29

**What was changed?**

For this issue, I updated the login page in addition to creating a forgot password page and reset password page. The user can now redirect themselves from the login page to the forgot password page. In the forgot password page, the user must enter their registered email account to receive a reset password link via email. The user can also navigate from the forgot password page to the login or signup page. The reset password link will be valid for 60 minutes or until the user accesses the link. The link will redirect the user to the reset password page. Here they must entire their new password twice, and the password will be validated before being updated in the database. Once the password is successfully updated, the user will be returned to the login page. I've also added automated testing for these new components.

**Why was it changed?**

My implementations of the forgot password page and reset password page, in addition to the modification of the login page, are needed to enhance user friendliness. Specifically, users may forget their password, and they need a chance to reset it securely so that they can access their account again. My code addresses this need by allowing users to receive a reset password link via email, which then allows users to enter a new valid password. Once their password is reset, they can login using it. 

**How was it changed?**

I had to update several files to address the needs of this issue. In App.js, I added routes to the forgot password and reset password pages. Within the LoginScreen.js, I modified the forgot password link so it redirects to the forgot password page upon clicking. I created ForgotPassword.js and ForgotPassword.css to make the forgot password page. ForgotPassword.js communicates the submitted user email with the backend. I also created ResetPassword.js and ResetPassword.css to make the reset password page. ResetPassword.js validates the user's password and communicates it with the backend. Server.js was updated so that it can validate the user's submitted email and then send a reset link to the user's email. Server.js also handles making sure the token a user is using to reset their password is valid and not expired. Server.js is also responsible for updating the user's password in the database. I also updated the README file to contain instructions on configuration details for the reset password functionality. Lastly, I updated LoginScreen.test.js and created ForgotPassword.test.js and ResetPassword.test.js to address UI interactions for the forgot password flow, backend validation and error handling, and a successful password reset scenario.

**Screenshots that show the changes (if applicable):**
image
image
image

### CODE REVIEW

**Review by AjithAkuthota23 — COMMENTED** [review]
1. The changes you made works good in light mode, but it does not completely visible in the dark mode. Could you change that part?
image

2. When I click on the reset link that I got in the mail, it redirects to the page where the UI components are not completely visible. 
image

3. I could successfully change the password

File: llm_bioinformatics_research/server.js [hunk 1]
```diff
@@ -166,6 +168,90 @@ app.post('/home', async (req, res) => {
...
+    const { token } = req.query;
+
+    if (!token) {
+        return res.status(400).send("Invalid or missing token.");
+    }
+
+    // Redirect to the frontend app with the token as a query parameter
+    res.redirect(`http://localhost:3000/reset-password?token=${token}`);
+});
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/server.js:254]: Password reset works successfully!

File: llm_bioinformatics_research/src/Components/ResetPassword/ResetPassword.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    setPasswordError(passwordValidationError);
+    if (passwordValidationError) return;
+  
+    if (newPassword !== confirmPassword) {
+      setConfirmPasswordError("Passwords do not match.");
+      return;
+    } else {
+      setConfirmPasswordError('');
+    }
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/src/Components/ResetPassword/ResetPassword.js:45]: In addition to the existing validation, I think we should implement a check that ensures that the new password is not the same as the existing password (we don't want the user to change it to the same password for security reasons).

File: llm_bioinformatics_research/src/Components/ForgotPassword/ForgotPassword.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      <p className="signUpText">
+        Need a new account? <a href="/signup" className="signUpLink">Sign up</a>
+      </p>
+      <p className="loginText">
+        Already have an account? <a href="/login" className="loginLink">Login</a>
+      </p>
+    </div>
+  );
+}
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/src/Components/ForgotPassword/ForgotPassword.js:62]: I think we should update the layout of this page to be similar to the login/sign up pages to ensure consistency across the application. Integrating the material UI library is necessary for those changes.

File: llm_bioinformatics_research/src/Components/ResetPassword/ResetPassword.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            className="reset-button"
+          >
+            Reset Password
+          </Button>
+        </form>
+      </Box>
+    </Container>
+  );
+}
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/src/Components/ResetPassword/ResetPassword.js:147]: Same thing for this page. I think we should update the layout of this page to be similar to the login/sign up pages to ensure consistency across the application. Integrating the material UI library is necessary for those changes.

File: llm_bioinformatics_research/src/Testing/ForgotPassword.test.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    const emailInput = screen.getByPlaceholderText("Email");
+    const submitButton = screen.getByText("Send email");
+  
+    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
+    fireEvent.click(submitButton);
+  
+    const expiredTokenMessage = await screen.findByText("Token expired. Please request a new password reset link.");
+    expect(expiredTokenMessage).toBeInTheDocument();
+  });
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/src/Testing/ForgotPassword.test.js:98]: For this test, how are we ensuring that the link has expired? Is it simply just the mocking the error 400? Can we somehow check the validity of the link by employing a timer to expiration? This is just a suggestion.

File: llm_bioinformatics_research/src/Testing/ResetPassword.test.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    const confirmPasswordInput = screen.getByLabelText("Confirm Password");
+    const resetButton = screen.getByRole("button", { name: /reset password/i });
+
+    userEvent.type(newPasswordInput, "Short1!");
+    userEvent.type(confirmPasswordInput, "Short1!");
+    fireEvent.click(resetButton);
+
+    expect(screen.getByText("Password must be at least 8 characters long.")).toBeInTheDocument();
+  });
+
```
Comment (PranaviKolouju) [inline llm_bioinformatics_research/src/Testing/ResetPassword.test.js:63]: As an addition to the comment I made earlier, we should a test to make sure that the new password is not the same as the existing password in the database. If it is, then the user should be prompted to modify the new password and we should add a test to check for this.

**Review by AjithAkuthota23 — APPROVED** [review]
The forgot password functionality is working good.

1. When an email is entered which is in the database , the email is sent to that email with the link associated to change the password.
2. After clicking the link, user can successfully change the password
3. The user cannot use the most recent password

Great work, everything is working as expected.

**Pull request merged**